### PR TITLE
Feature/rest producer

### PR DESCRIPTION
**Pull request by masud-svg**
**REST Producer**

### CODE REVIEW

File: databridge.examples/databridge.examples.aas-jsonata-http/src/test/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/test/TestAASUpdater.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+		
+		aasServer.stopComponent();
+	}
+	
+	@Test
+	public void CheckPropertyValueA() throws  InterruptedException, ClientProtocolException, IOException {
+		
+		String expected_value = "103.5585973";
+		
+		startHTTPServer(getResponseBody(expected_value));
```
Comment (mdanish98) [inline databridge.examples/databridge.examples.aas-jsonata-http/src/test/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/test/TestAASUpdater.java:102]: This here sets the value to the mocked server, and while retrieving returns the same value, as explained in the above comment.

**Review by mdanish98 — CHANGES_REQUESTED** [review]
I have added some of the initial review remarks. I will review it again after the current remarks are addressed.

**One general remark:**
This PR is missing a regression test for the DataBridge component.

The exact location for test addition: [DataBridge Component Test](https://github.com/eclipse-basyx/basyx-databridge/tree/main/databridge.component/src/test/java/org/eclipse/digitaltwin/basyx/databridge/executable/regression)

For reference, please take a look at the already existing regression tests there, such as MQTT, Plc4X, etc.
Please follow the same structure and design as in other regression tests.

File: databridge.camel-httppolling/src/main/java/org/eclipse/digitaltwin/basyx/databridge/httppolling/configuration/HttpProducerConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	}
+
+	public void setApi(String api) {
+		this.api = api;
+	}
+
+	@Override
+	public String getConnectionURI() {
+		
+		StringBuilder endpointDefinition = new StringBuilder();
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.camel-httppolling/src/main/java/org/eclipse/digitaltwin/basyx/databridge/httppolling/configuration/HttpProducerConfiguration.java:72]: updated

File: databridge.camel-httppolling/src/main/java/org/eclipse/digitaltwin/basyx/databridge/httppolling/configuration/factory/HttpProducerDefaultConfigurationFactory.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ ******************************************************************************/
+
+package org.eclipse.digitaltwin.basyx.databridge.httppolling.configuration.factory;
+
+import org.eclipse.digitaltwin.basyx.databridge.core.configuration.factory.DataSinkConfigurationFactory;
+import org.eclipse.digitaltwin.basyx.databridge.httppolling.configuration.HttpProducerConfiguration;
+
+/**
+ * A default configuration factory for HTTP Producer
+ * @author rana
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.camel-httppolling/src/main/java/org/eclipse/digitaltwin/basyx/databridge/httppolling/configuration/factory/HttpProducerDefaultConfigurationFactory.java:33]: changed

File: databridge.examples/databridge.examples.aas-jsonata-http/src/test/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/test/TestAASUpdater.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+	private static DataBridgeComponent updater;
+	private static int PORT = 8080;
+	private static ClientAndServer clientServer;
+	
+	@BeforeClass
+	public static void setUp() throws Exception {
+
+		configureAndStartAasServer();
+		
+		mockServerConfig();
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.examples/databridge.examples.aas-jsonata-http/src/test/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/test/TestAASUpdater.java:85]: changed

File: databridge.examples/databridge.examples.aas-jsonata-http/src/test/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/test/TestAASUpdater.java (hunk 1 above)
Comment (masud-svg) [reply databridge.examples/databridge.examples.aas-jsonata-http/src/test/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/test/TestAASUpdater.java:102]: changed

File: databridge.examples/databridge.examples.aas-jsonata-http/src/test/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/test/TestAASUpdater.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+		
+		assertEquals(actualValue, expected_value);
+	}
+	
+	@Test
+	public void CheckPropertyValueB() throws  InterruptedException, ClientProtocolException, IOException, URISyntaxException {
+		
+		String expected_value = getExpectedValueFromFile();
+		
+		startHTTPServer(getResponseBody(expected_value));
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.examples/databridge.examples.aas-jsonata-http/src/test/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/test/TestAASUpdater.java:115]: changed

File: databridge.examples/databridge.examples.aas-jsonata-http/src/main/resources/routes.json
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.examples/databridge.examples.aas-jsonata-http/src/main/resources/routes.json:1]: In our current context, we are using AAS Polling consumer as the source that's why it's needed.

**Review by mdanish98 — CHANGES_REQUESTED** [review]
Following are some of the review remarks.

While executing the regression test of databridge component, I am getting NoHttpResponseException. I have attached logs for this.

Please take a look and fix this issue.
[NoHttpResponseException.log](https://github.com/eclipse-basyx/basyx-databridge/files/14053246/NoHttpResponseException.log)

File: databridge.component/src/test/java/org/eclipse/digitaltwin/basyx/databridge/executable/regression/DataBridgeSuiteHttpProducer.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.eclipse.basyx.aas.aggregator.api.IAASAggregator;
+import org.eclipse.digitaltwin.basyx.databridge.examples.aasjsonatahttp.Server;
+import org.eclipse.digitaltwin.basyx.databridge.examples.httpserver.HttpDataSource;
+import org.junit.Test;
+
+/**
+ * Suite for testing that the DataBridge is setup correctly with HTTPP Producer
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.component/src/test/java/org/eclipse/digitaltwin/basyx/databridge/executable/regression/DataBridgeSuiteHttpProducer.java]: changed

File: databridge.component/src/test/java/org/eclipse/digitaltwin/basyx/databridge/executable/regression/DataBridgeSuiteHttpProducer.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+ *
+ */
+public abstract class DataBridgeSuiteHttpProducer {
+	
+	protected abstract IAASAggregator getAASAggregatorProxy();
+	protected abstract String getEndpoint();
+	protected abstract String getHost();
+	private static HttpDataSource httpData;
+	private static HttpServlet httpServlet = new Server();
+	private static int PORT = 8091;
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.component/src/test/java/org/eclipse/digitaltwin/basyx/databridge/executable/regression/DataBridgeSuiteHttpProducer.java]: chnaged

File: databridge.component/src/test/java/org/eclipse/digitaltwin/basyx/databridge/executable/regression/TestDataBridgeHttpProducer.java
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.component/src/test/java/org/eclipse/digitaltwin/basyx/databridge/executable/regression/TestDataBridgeHttpProducer.java:1]: chnaged

File: databridge.component/src/test/java/org/eclipse/digitaltwin/basyx/databridge/executable/regression/TestDataBridgeHttpProducer.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+		aasServer.setRegistry(registry);
+		aasServer.startComponent();
+	}
+	
+	private static void startUpdaterComponent() {
+		DataBridgeExecutable.main(new String[] {"src/test/resources/httpproducer/databridge"});
+	}
+	
+	private static void stopDataBridgeComponent() {
+		if(DataBridgeExecutable.getDataBridgeComponent() != null) {
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.component/src/test/java/org/eclipse/digitaltwin/basyx/databridge/executable/regression/TestDataBridgeHttpProducer.java]: changed

File: databridge.component/src/test/resources/httpproducer/databridge/httpproducer.json [hunk 9]
```diff
@@ -0,0 +1,9 @@
+[
+	{
+		"uniqueId": "Server1",
+		"scheme": "http",
+		"ServerUrl": "localhost",
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.component/src/test/resources/httpproducer/databridge/httpproducer.json]: chnaged

File: databridge.examples/databridge.examples.aas-jsonata-http/src/main/resources/httpproducer.json [hunk 10]
```diff
@@ -0,0 +1,16 @@
+[
+	{
+		"uniqueId": "Server1",
+		"scheme": "http",
+		"ServerUrl": "localhost",
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.examples/databridge.examples.aas-jsonata-http/src/main/resources/httpproducer.json]: changed

File: databridge.examples/databridge.examples.aas-jsonata-http/src/main/resources/httpproducer.json [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+		"path": "data/machine/pressure",
+		"port": 8091
+	},
+	{
+		"uniqueId": "Server2",
+		"scheme": "http",
+		"ServerUrl": "localhost",
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.examples/databridge.examples.aas-jsonata-http/src/main/resources/httpproducer.json]: changed

File: databridge.examples/databridge.examples.aas-jsonata-http/src/main/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/Server.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+	
+	/**
+	 * Sets retrieved message as a response
+	 * @param response
+	 * @param resp
+	 * @throws IOException
+	 */
+	private void setMessageToResponse(HttpServletResponse resp) throws IOException {
+
+		String jsonString = requestBodyValue;
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.examples/databridge.examples.aas-jsonata-http/src/main/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/Server.java]: changed

File: databridge.examples/databridge.examples.aas-jsonata-http/src/main/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/Server.java
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.examples/databridge.examples.aas-jsonata-http/src/main/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/Server.java:1]: changed

File: databridge.camel-httppolling/src/main/java/org/eclipse/digitaltwin/basyx/databridge/httppolling/configuration/HttpProducerConfiguration.java [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+/**
+ * An implementation of HTTP producer configuration
+ * 
+ * @author rana
+ *
+ */
+public class HttpProducerConfiguration extends DataSinkConfiguration{
+	
+	private String scheme;
+	private String ServerUrl;
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.camel-httppolling/src/main/java/org/eclipse/digitaltwin/basyx/databridge/httppolling/configuration/HttpProducerConfiguration.java]: changed

File: databridge.examples/databridge.examples.aas-jsonata-http/src/test/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/test/TestAASUpdater.java [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+		
+		String actualValue = getContentFromEndPoint();
+		
+		assertEquals(expected_value, actualValue);
+	}
+	
+	@Test
+	public void CheckPropertyValueB() throws  InterruptedException, IOException, URISyntaxException {
+		
+		String expected_value = getExpectedValueFromFile();
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.examples/databridge.examples.aas-jsonata-http/src/test/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/test/TestAASUpdater.java:112]: changed

File: databridge.examples/databridge.examples.httpserver/src/main/java/org/eclipse/digitaltwin/basyx/databridge/examples/httpserver/DummyServlet.java [hunk 15]
```diff
@@ -27,12 +27,18 @@
...
+import javax.servlet.ServletException;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
 public class DummyServlet extends HttpServlet {
 	private static final long serialVersionUID = 4918478763760299634L;
+	
+	@Override
+	protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.examples/databridge.examples.httpserver/src/main/java/org/eclipse/digitaltwin/basyx/databridge/examples/httpserver/DummyServlet.java]: changed

File: databridge.examples/databridge.examples.httpserver/src/main/java/org/eclipse/digitaltwin/basyx/databridge/examples/httpserver/HttpDataSource.java [hunk 16]
```diff
@@ -33,6 +35,12 @@ public void runHttpServer() throws InterruptedException {
 		server.start();
 	}
 	
+	public void runHttpServer(String host, int port, HttpServlet httpServlet) throws InterruptedException {
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.examples/databridge.examples.httpserver/src/main/java/org/eclipse/digitaltwin/basyx/databridge/examples/httpserver/HttpDataSource.java]: chnaged

File: databridge.examples/databridge.examples.aas-jsonata-http/src/test/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/test/TestAASUpdater.java [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+	private static HttpDataSource httpData;
+	private static HttpServlet httpServlet = new Server();
+	private static int PORT = 8091;
+
+	@BeforeClass
+	public static void setUp() throws Exception {
+
+		configureAndStartAasServer();
+		
+		serverStart();
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.examples/databridge.examples.aas-jsonata-http/src/test/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/test/TestAASUpdater.java]: changed

File: databridge.examples/databridge.examples.aas-jsonata-http/src/test/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/test/TestAASUpdater.java [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+		
+		serverStop();
+	}
+	
+	@Test
+	public void CheckPropertyValueA() throws  InterruptedException, IOException {
+		
+		String expected_value = wrapStringValue("103.5585973");
+		
+		Thread.sleep(10000);
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.examples/databridge.examples.aas-jsonata-http/src/test/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/test/TestAASUpdater.java]: changed

File: databridge.examples/databridge.examples.aas-jsonata-http/src/test/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/test/TestAASUpdater.java [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+		
+		aasServer.stopComponent();
+		
+		serverStop();
+	}
+	
+	@Test
+	public void CheckPropertyValueA() throws  InterruptedException, IOException {
+		
+		String expected_value = wrapStringValue("103.5585973");
```
[the remark this answers is not in the document]
Comment (masud-svg) [reply databridge.examples/databridge.examples.aas-jsonata-http/src/test/java/org/eclipse/digitaltwin/basyx/databridge/examples/aasjsonatahttp/test/TestAASUpdater.java:101]: changed

**Review by mdanish98 — APPROVED** [review]
Approved

**Pull request merged**